lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -0,0 +1,313 @@
+
+use bits::*;
+use cpuid_utils::{CpuidIdent, CpuidSet, CpuidValues};
+use hypercall::{hypercall_page_contents, MsrHypercallValue};
+use slog::info;
+
+use crate::{
+    accessors::MemAccessor,
+    common::{GuestRegion, Lifecycle, VcpuId, PAGE_SIZE},
+    enlightenment::AddCpuidError,
+    migrate::{
+        MigrateCtx, MigrateSingle, MigrateStateError, Migrator, PayloadOffer,
+        PayloadOutput,
+    },
+    msr::{MsrId, RdmsrOutcome, WrmsrOutcome},
+    vmm::SubMapping,
+};
+
+mod bits;
+mod hypercall;
+
+const TYPE_NAME: &str = "guest-hyperv-interface";
+
+#[derive(Debug, Default)]
+struct Inner {
+    /// The last value stored in the [`bits::HV_X64_MSR_GUEST_OS_ID`] MSR.
+    msr_guest_os_id_value: u64,
+
+    /// The last value stored in the [`bits::HV_X64_MSR_HYPERCALL`] MSR.
+    msr_hypercall_value: MsrHypercallValue,
+}
+
+pub struct HyperV {
+    log: slog::Logger,
+    inner: Mutex<Inner>,
+    acc_mem: MemAccessor,
+}
+
+impl HyperV {
+    /// Creates a new Hyper-V enlightenment stack.
+    pub fn new(log: &slog::Logger) -> Self {
+        let acc_mem = MemAccessor::new_orphan();
+        let log = log.new(slog::o!("component" => "hyperv"));
+        Self { log, inner: Mutex::new(Inner::default()), acc_mem }
+    }
+
+    /// Handles a write to the HV_X64_MSR_GUEST_OS_ID register.
+    fn handle_wrmsr_guest_os_id(&self, value: u64) -> WrmsrOutcome {
+        let mut inner = self.inner.lock().unwrap();
+
+        // TLFS section 3.13 says that the hypercall page "becomes disabled" if
+        // the guest OS ID register is cleared after the hypercall register is
+        // set. It also specifies that attempts to set the Enabled bit in that
+        // register will be ignored if the guest OS ID is zeroed, so handle this
+        // case by clearing the hypercall MSR's Enabled bit but otherwise
+        // leaving the hypercall page untouched (as would happen if the guest
+        // manually cleared this bit).
+        if value == 0 {
+            info!(&self.log, "guest cleared HV_X64_MSR_GUEST_OS_ID");
+            inner.msr_hypercall_value.clear_enabled();
+        } else {
+            info!(
+                &self.log,
+                "guest set HV_X64_MSR_GUEST_OS_ID";
+                "value" => value
+            );
+        }
+
+        inner.msr_guest_os_id_value = value;
+        WrmsrOutcome::Handled
+    }
+
+    /// Handles a write to the HV_X64_MSR_HYPERCALL register. See TLFS section
+    /// 3.13.
+    fn handle_wrmsr_hypercall(&self, value: u64) -> WrmsrOutcome {
+        let mut inner = self.inner.lock().unwrap();
+        let old = inner.msr_hypercall_value;
+
+        // Per the spec, this MSR is immutable once the Locked bit is set.
+        if old.locked() {
+            return WrmsrOutcome::Handled;
+        }
+
+        // If this MSR is written when no guest OS ID is set, the Enabled bit is
+        // cleared and the write succeeds.
+        let mut new = MsrHypercallValue(value);
```

Comment:
we're taking the guest's provided bits 2 through 11 here, which don't get used but i'm not sure we really _want_ to retain? i suppose the TLFS doesn't say anything other than guests _should_ preserve these bits, and #GP if any of the reserved bits are touched might be too dramatic.

Reply:
Added a note about these bits in 5bf277f7.